cc/main.cc

```diff
@@ -465,46 +468,232 @@ class SelfPlayer {
+
+    for (int i = 0; i < num_games;) {
+      threads_.emplace_back(std::bind(&Evaluator::ThreadRun, this, i,
+                                      cur_model.get(), prev_model.get()));
+      ++i;
+      if (FLAGS_even && i < num_games) {
+        threads_.emplace_back([&, i] {
+          // Wait for the other games to play the first move.
+          barrier_->Wait();
+          ThreadRun(i, prev_model.get(), cur_model.get());
+        });
+        ++i;
+      }
+    }
+
+    for (auto& t : threads_) {
+      t.join();
+    }
+
+    std::cerr << "Evaluated " << num_games << " games, total time "
+              << absl::ToDoubleSeconds(absl::Now() - start_time) << " sec."
+              << std::endl;
+
+    auto print_name = [](std::string name) {
+      if (name.size() > 20) {
+        name.resize(17);
+        name.append("...");
+      } else {
+        name.append(20 - name.size(), ' ');
+      }
+      std::cerr << name;
+    };
+    auto print_wins = [&](int wins) {
+      std::cerr << "   " << std::setw(3) << wins << " " << std::setw(6)
+                << wins * 100.0f / num_games << "%";
+    };
+    auto print_result = [&](const Model& model) {
+      print_name(model.name);
+      print_wins(model.black_wins + model.white_wins);
+      print_wins(model.black_wins);
+      print_wins(model.white_wins);
+      std::cerr << std::endl;
+    };
+
+    std::cerr << "wins                        total         black         white"
+              << std::endl
+              << std::fixed << std::setprecision(2);
+    print_result(*prev_model);
+    print_result(*cur_model);
+    std::cerr << "                                  ";
+    print_wins(prev_model->black_wins + cur_model->black_wins);
+    print_wins(prev_model->white_wins + cur_model->white_wins);
+    std::cerr << std::endl;
+  }
+
+ private:
+  void ThreadRun(int thread_id, Model* model, Model* other_model) {
+    // The player and other_player reference this pointer.
+    std::unique_ptr<DualNet> dual_net;
+
+    auto player_options = options_;
+    // If an random seed was explicitly specified, make sure we use a
+    // different seed for each thread.
+    if (player_options.random_seed != 0) {
+      player_options.random_seed += 1299283 * thread_id;
+    }
+
+    player_options.verbose = thread_id == 0;
+    player_options.name = model->name;
+    auto player = absl::make_unique<MctsPlayer>(
+        absl::make_unique<WrappedDualNet>(dual_net), player_options);
+
+    player_options.verbose = false;
+    player_options.name = other_model->name;
+    auto other_player = absl::make_unique<MctsPlayer>(
+        absl::make_unique<WrappedDualNet>(dual_net), player_options);
+
+    auto* black = player.get();
+    auto* white = other_player.get();
+
+    auto* factory = model->factory.get();
+    auto* other_factory = other_model->factory.get();
+
+    while (!player->game_over()) {
+      dual_net = factory->New();
+      barrier_->Wait();  // Wait for all threads to create DualNet.
```

Comment:
sorry, i'm not sure i understand why this is needed?

Reply:
I've added additional comments to hopefully explain why this is needed.